webui/react/src/stores/userSettings.tsx

```diff
@@ -141,8 +145,42 @@ class UserSettingsStore extends PollingStore {
     }
   }
 
+  /**
+   * This updates the value of a setting and persists it for future sessions.
+   * @param type The type of the value or an encoder of the value to JSON.
+   * @param key Unique key to store and retrieve the settings
+   * @param fn Function to update the value of the setting at `key`
+   */
+  public update<T extends t.Props, U = t.TypeOfProps<T>>(
+    type: t.TypeC<T>,
+    key: string,
+    fn: (value: U | undefined) => U,
+  ): void;
+  public update<T>(type: t.Type<T, Json>, key: string, fn: (value: T | undefined) => T): void {
+    this.#settings.update((settings) => {
+      return Loadable.map(settings, (settings) => {
+        return settings.update(key, (jsonValue) => {
+          let value: T | undefined = undefined;
+          if (jsonValue !== undefined) {
+            const attempt = type.decode(jsonValue);
+            if (isRight(attempt)) {
+              value = attempt.right;
+            }
+          }
+          const newValue = fn(value);
+          // This is non-blocking. If the API call fails we don't want to block
+          // the user from interacting, just let them know that their settings
+          // are not persisting. It's also important to update the value immediately
+          // for good rendering performance.
+          setTimeout(() => this.updateUserSetting(key, newValue), 0);
```

Comment:
Is there a underlying lock, I wonder would this run into some race conditions?

Reply:
So this part is just the API call. It's possible this isn't the correct way to do this but I wanted to explicitly move the DB call outside of the observable update for `update` since we want that to be a fast local update. In order to make sure the sequencing is correct when calling `set` and `update` both in the same scheduler task I changed `set` to also bump out of the current task with a timeout.

My understanding is that this should schedule these updates in the order the functions were called. That said, we shouldn't be doing making multiple calls to update the same setting in the same scheduler task _anyway_. `userSettings.set(T, k, foo); userSettings.set(T, k, bar)` is obviously incorrect code. Under a reactive programming paradigm if there's a dependency between actions you should model that using a subscription (`useEffect`).

_That_ said it is still technically possible—even when properly modeled—for the db updates to occur out of order due to network issues. I think technically to guarantee correct ordering we should be putting DB updates on a queue and issuing them synchronously. We'd probably also want some kind of error handling mechanism for when a DB update fails but there are updates queued behind them?

Reply:
Thanks for the explanation, network issue is something to consider, but I also want to ask about local update. I believe the reason we use `update(prev=>prev+1)` over `set(state+1)` is to prevent concurrency, if we just retrieve the previous value in `update`, isn't that just the same as `set(state+1)`?

Reply:
(In case it's not obvious `this.updateUserSetting` is just the API part of the update)

Reply:
I think you're correct here. I'll just remove the timeouts.